### ISSUE TITLE
Automatically prune the global store

### ISSUE THREAD

**Issue by zkochan** [pnpm/pnpm#2694]
Currently, pnpm never prunes the storage. So if some packages are not used anymore, they will stay in the store until `pnpm store prune` is executed.

We need to prune the store automatically but not each time the command runs. It is probably enough to prune the store when `pnpm install` is executed but only once a week. Alternatively, we could print a message suggesting to run `pnpm store prune`

Edit:
It might make sense to only prune the store if the size of the store has grown more than 10% or some other threshold

**Comment by nikoladev** [comment]
Slightly related to this: I recently noticed that pruning the global store only works for the current pnpm version. Anything that was installed with a previous version was not pruned. I had to install `pnpm@4` to prune the older global store.

Is it possible to also ensure the prune command also prunes the global store created by previous versions of pnpm? If not, maybe we could add a link to some documentation that explains the workaround I described above?

**Comment by zkochan** [comment]
How should it prune the previous versions? Just remove the whole store of the previous version?

**Comment by nikoladev** [comment]
Is it not possible to use the current pnpm major version to see which packages are not used if those packages were installed by a previous pnpm version?

Would it be possible to temporarily install the old version of pnpm (the one that was used for those packages) and use that to prune the global store?

(No idea how doable this is, btw. I'm just brainstorming here)

**Comment by nikoladev** [comment]
Another suggestion: currently the user doesn't know how much space was saved by pruning, only the amount of files and packages. Would it be possible to also show the amount of space saved? Because that is something that I at least want to know when I prune the store: how much space has this saved me?

**Comment by zkochan** [comment]
> Is it not possible to use the current pnpm major version to see which packages are not used if those packages were installed by a previous pnpm version?

No, that code is removed from the last pnpm. We can print a message that another pnpm version's store is present. If someone upgraded pnpm, they can remove the olde store. If they still need to use the old version, they should use the old version for prune.

> Another suggestion: currently the user doesn't know how much space was saved by pruning, only the amount of files and packages. Would it be possible to also show the amount of space saved? Because that is something that I at least want to know when I prune the store: how much space has this saved me?

Maybe we can do that but it will probably need more filesystem operations to be performed.

**Comment by nikoladev** [comment]
> We can print a message that another pnpm version's store is present

That's probably the easiest acceptable solution then.

> Maybe we can do that but it will probably need more filesystem operations to be performed.

It's not super important, so maybe it can be put on the backlog for when we add some other functionality that also needs the same filesystem operations?

**Comment by probablykasper** [comment]
Having to manually prune the store sounds pretty unnecessary imo, I'm probably holding off on switching to pnpm because of this

**Comment by zkochan** [comment]
This is a weird reason to not use pnpm. I personally never prune the store because it remains so small.

**Comment by probablykasper** [comment]
Not saying it's a massive issue, but it's an extra thing to think about that I would rather not, even if that's weird.

**Comment by zkochan** [comment]
This for sure will be implemented but it is not a big deal even if you don't clear the store ever. Most of the files don't change even when a new version of the package is published, so the store growth really slowly. My global content-addressable store is currently only 3GB and I have a lot of projects on my disk.

**Comment by Bananaman** [comment]
I like the idea of keeping track of the store-size when it was last pruned, and after a 10% growth of the library, it recommends a prune and then records the result as the new size. Very good idea.

The issue I see is how to calculate the size of the storage quickly. Is it worth checking the size every time the user installs a package? Or just a certain time after the last check.

- Keep track of size when store was last pruned. Only recommend prune when it has grown by 10%+.
- Keep track of timestamp when `install` last checked the store-size. Only check it again once per day at most.

How does this sound?
